src/compiler/compiler.rs

```diff
@@ -48,6 +48,62 @@ use tempfile::TempDir;
 
 use crate::errors::*;
 
+#[macro_use]
+pub mod util {
+
+    // Helper macro to create static array of calculable fixed length
+    #[macro_export]
+    macro_rules! counted_array {
+        // INTERNAL
+
+        // last element (proceed to output)
+        (@parse $size:expr, ($val:expr) -> [$($accs:expr),*] $thru:tt) => {
+            counted_array!(@output $size + 1usize, [$($accs,)* $val] $thru);
+        };
+        // more elements (keep parsing)
+        (@parse $size:expr, ($val:expr, $($vals:expr),*) -> [$($accs:expr),*] $thru:tt) => {
+            counted_array!(@parse $size + 1usize, ($($vals),*) -> [$($accs,)* $val] $thru);
+        };
+
+        // output a local variable
+        (@output $size:expr, $acc:tt (() let $n:ident $t:ty)) => {
+            let $n: [$t; $size] = $acc;
+        };
+        // output a lazy static
+        (@output $size:expr, $acc:tt (($($p:tt)*) lazy_static $n:ident $t:ty)) => {
+            lazy_static!{ $($p)* static ref $n: [$t; $size] = $acc; }
+        };
+        // output a static or const item
+        (@output $size:expr, $acc:tt (($($p:tt)*) $s:ident $n:ident $t:ty)) => {
+            $($p)* $s $n: [$t; $size] = $acc;
+        };
+
+        // EXTERNAL
+
+        // entry point
+        (pub $storage:ident $n:ident: [$t:ty; _] = [$($vals:expr),* $(,)*]) => {
+            counted_array!(@parse 0usize, ($($vals),*) -> [] ((pub) $storage $n $t));
+        };
+        (pub $restr:tt $storage:ident $n:ident: [$t:ty; _] = [$($vals:expr),* $(,)*]) => {
+            counted_array!(@parse 0usize, ($($vals),*) -> [] ((pub $restr) $storage $n $t));
+        };
+        ($storage:ident $n:ident: [$t:ty; _] = [$($vals:expr),* $(,)*]) => {
+            counted_array!(@parse 0usize, ($($vals),*) -> [] (() $storage $n $t));
+        };
+    }
+}
+
+#[cfg(test)]
+mod util_test {
+    use crate::counted_array;
+
+    #[test]
+    fn counted_array_macro() {
+        counted_array!(static ARR_QUAD: [u8;_] = [1,2,3,4,]);
+        assert_eq!(ARR_QUAD.len(), 4);
+    }
+}
```

Comment:
I'd prefer this to be in a separate file.

Is there a particular reason we go for the entire macro rather than i.e. using the 11 liner I wrote
when that's really all we need?
Half of it is dead code iiuc and we infer the maintenance burdon.

Reply:
I couldn't get the macro you wrote to work reliably, it was having trouble with scope. Macros aren't my strongest area so I just went with the path of least resistance. I was super tired so I didn't realize there were dead code areas in the macro, I'll clean those up and move it to a separate file.

Reply:
Btw, thank you for being patient. My fixes aren't as smooth as they could be because I don't have a ton of Rust experience.